File: entities/particle.py
import math
import random
import pygame

from config import *
from utils.color import *


class Particle:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.vx = random.uniform(-0.5, 0.5)
        self.vy = random.uniform(-0.5, 0.5)
        self.speed = 0
        self.max_speed = MAX_VELOCITY
        self.size = random.choice(PARTICLE_SIZE)

    def update(self, black_hole):
        dx = black_hole.x - self.x
        dy = black_hole.y - self.y
        dist = math.sqrt(dx**2 + dy**2)
        
        force = GRAVITATION_CONSTANT / (dist**FIELD_VARIATION)   # gravity

        if dist > black_hole.radius:
            angle = math.atan2(dy, dx)
            
            # Tangential components
            tangential_force = force * TANGENTIAL_FACTOR
            self.vx += force * math.cos(angle) - tangential_force * math.sin(angle)
            self.vy += force * math.sin(angle) + tangential_force * math.cos(angle)
            
            # Limiting velocity
            self.speed = math.sqrt(self.vx**2 + self.vy**2)
            speed=self.speed

            max_speed = self.max_speed

            if speed > max_speed:
                self.vx = self.vx / speed * max_speed
                self.vy = self.vy / speed * max_speed
        
        self.x += self.vx
        self.y += self.vy

        self.force=force
        

    def draw(self, screen, black_hole):
        
        if (COLOR_VARIATION == 1):
            color = get_color_based_on_force(self.force, 3)
        elif (COLOR_VARIATION == 2):
            color = get_color_based_on_speed(self.speed, self.max_speed) 
        else:
            color = WHITE

        dist = math.sqrt((self.x - black_hole.x)**2 + (self.y - black_hole.y)**2)
        
        if 0 <= self.x < WIDTH and 0 <= self.y < HEIGHT:
            pygame.draw.circle(screen, color, (int(self.x), int(self.y)), self.size)

File: src/Particle.cpp
#include "Particle.h"
#include "BlackHole.h"
#include "config.h"
#include <cmath>
#include <random>

// Helper for random number generation
namespace {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(-0.5, 0.5);
}

const float Particle::MAX_SPEED = MAX_VELOCITY;

Particle::Particle(float x, float y) : pos(x, y), speed(0), force(0) {
    vel.x = dis(gen);
    vel.y = dis(gen);
}

void Particle::update(const BlackHole& blackHole) {
    sf::Vector2f delta = blackHole.pos - this->pos;
    float dist = std::sqrt(delta.x * delta.x + delta.y * delta.y);

    if (dist > blackHole.radius) {
        // Gravitational force
        this->force = GRAVITATION_CONSTANT / std::pow(dist, FIELD_VARIATION);

        float angle = std::atan2(delta.y, delta.x);

        // Calculate gravitational and tangential acceleration
        sf::Vector2f acceleration;
        acceleration.x = this->force * std::cos(angle);
        acceleration.y = this->force * std::sin(angle);

        sf::Vector2f tangential_accel;
        tangential_accel.x = -this->force * TANGENTIAL_FACTOR * std::sin(angle);
        tangential_accel.y = this->force * TANGENTIAL_FACTOR * std::cos(angle);

        // Update velocity
        this->vel += acceleration + tangential_accel;

        // Limit velocity
        this->speed = std::sqrt(this->vel.x * this->vel.x + this->vel.y * this->vel.y);
        if (this->speed > MAX_SPEED) {
            this->vel = (this->vel / this->speed) * MAX_SPEED;
            this->speed = MAX_SPEED; // Update speed after clamping
        }
    } else {
        // To avoid division by zero or extreme forces if a particle gets too close
        this->force = 0;
        this->speed = 0;
    }

    // Update position
    this->pos += this->vel;
}
